### ISSUE TITLE
siamese adaptation?

### ISSUE THREAD

**Issue by phobrain** [kutoga/going_deeper#2]
Since an Input layer 0 is baked in: ...ginfty.py", line 331, in init
    assert isinstance(self._layers[0], Input((1,)).__class__)
AssertionError

I can't simply plug into this model, and perhaps other downstream issues will be obvious. I'm not sure if that's an easy fix, or if it might be more appropriate to take a different approach than siamese, that being the only one I've used so far.

    w_step = 5
    f_reg = 1e-8
    w_reg = 5e-5
    dense_size = 175
    base_network = TDModel()
    base_network += GInftlyLayer(
            'dfc0',
            w_regularizer=(c_l2, 1e-3),
            f_regularizer=(c_l2, f_reg),
            reweight_regularizer=False,
            f_layer=[
                lambda reg: Dense(dense_size),
                lambda reg: Dropout(0.1),
                lambda reg: GammaRegularizedBatchNorm(reg, max_free_gamma=0.),
            ], h_step=[
                lambda reg: Activation('relu'),
            ],
            w_step=w_step,
        )
        base_network += Dense(dense_size/4)
        base_network += GInftlyLayer(
            'dfc1',
            w_regularizer=(c_l2, 1e-3),
            f_regularizer=(c_l2, f_reg),
            reweight_regularizer=False,
            f_layer=[
                lambda reg: Dense(dense_size),
                lambda reg: Dropout(0.1),
                lambda reg: GammaRegularizedBatchNorm(reg, max_free_gamma=0.),
            ], h_step=[
                lambda reg: Activation('relu'),
            ],
            w_step=w_step,
        )
        # not sure how this factors into Model setup, below
        base_network.init(
            optimizer='adadelta', # TODO: adadelta needs to store the state; that is quite tricky, I think...
            loss='categorical_crossentropy',
            metrics=['categorical_accuracy']
        )

    input_a = Input(input_shape)
    input_b = Input(input_shape)

    processed_a = base_network(input_a)
    processed_b = base_network(input_b)

    distance = Lambda(poincare_distance,
                  output_shape=poinc_dist_output_shape)([processed_a, processed_b])

    model = Model([input_a, input_b], distance)
    model.compile(loss=contrastive_loss, optimizer=RMSprop())

**Comment by kutoga** [comment]
Unfortunately, the model is not that nice integrated in Keras. It is more the other way around: Keras is integrated into this model (which is worse than the other option).

First, I've tried to create a "real" Keras layer that wraps these ideas, but this was quite hard and I failed. I don't know how I can allocate more weights during the training in a standard Keras model. Therefore, I just wrote this very simple wrapper model that allows to play around with sequential models. I used Keras as "backend", because it allows to make simple tests.

One could generalize this architecture to allow multiple inputs (like in your example), but this is quite a lot of work. In general it would be better if someone could integrate this Layer-type in Keras, but for this I do not have enough knowledge.

Unfortunately, currently there are only sequential models possible with this code. A very simple alternative is to just use a normal Keras model and integrate this "GInftyLayer" in a static way: Just allow a maximum layer count of, e.g., 20 and allocate them statically (just add in a loop all the layers manually). This should work fine, but of course with the disadvantages that the training is slower (because all layers are allocated from the beginning) and there are not more than 20 sub-layer possible.

I'm sorry, that I cannot give you a better answer or change the code; The code is currently anyway not ready for production use, but only for testing purposes.

Thanks for your feedback:)!

**Comment by phobrain** [comment]
No problem, I can try a inputC = concat inputA, inputB approach, which I've been thinking of for asymmetric matches anyway. 

I guess one problem with implementing stuff in keras is you have to implement in each underlying platform.

**Comment by kutoga** [comment]
Things that can be implemented with Keras' (quite nice) Backend-interface do not need code for each underlying plattform and in this case I think this should work. But I think this type of meta-layer is currently not that easy implementable in Keras (I am happy if somehow can show that I'm wrong^^).
